src/node/http.ts

```diff
@@ -1,885 +1,253 @@
+ */
+export const redirect = (
+  req: express.Request,
+  res: express.Response,
+  to: string,
+  override: expressCore.Query = {},
+): void => {
+  const query = Object.assign({}, req.query, override)
+  Object.keys(override).forEach((key) => {
+    if (typeof override[key] === "undefined") {
+      delete query[key]
+    }
+  })
+
+  const relativePath = normalize(`${relativeRoot(req)}/${to}`, true)
+  const queryString = qs.stringify(query)
+  const redirectPath = `${relativePath}${queryString ? `?${queryString}` : ""}`
+  logger.debug(`redirecting from ${req.originalUrl} to ${redirectPath}`)
+  res.redirect(redirectPath)
 }
 
 /**
- * An HTTP server. Its main role is to route incoming HTTP requests to the
- * appropriate provider for that endpoint then write out the response. It also
- * covers some common use cases like redirects and caching.
+ * Get the value that should be used for setting a cookie domain. This will
+ * allow the user to authenticate only once. This will use the highest level
+ * domain (e.g. `coder.com` over `test.coder.com` if both are specified).
  */
-export class HttpServer {
-  protected readonly server: http.Server | https.Server
-  private listenPromise: Promise<string | null> | undefined
-  public readonly protocol: "http" | "https"
-  private readonly providers = new Map<string, HttpProvider>()
-  public readonly heart: Heart
-  private readonly socketProvider = new SocketProxyProvider()
-
-  /**
-   * Provides the actual proxying functionality.
-   */
-  private readonly proxy = proxy.createProxyServer({})
-
-  public constructor(private readonly options: HttpServerOptions) {
-    this.heart = new Heart(path.join(paths.data, "heartbeat"), async () => {
-      const connections = await this.getConnections()
-      logger.trace(plural(connections, `${connections} active connection`))
-      return connections !== 0
-    })
-    this.protocol = this.options.cert ? "https" : "http"
-    if (this.protocol === "https") {
-      this.server = httpolyglot.createServer(
-        {
-          cert: this.options.cert && fs.readFileSync(this.options.cert),
-          key: this.options.certKey && fs.readFileSync(this.options.certKey),
-        },
-        this.onRequest,
-      )
-    } else {
-      this.server = http.createServer(this.onRequest)
-    }
-    this.proxy.on("error", (error, _request, response) => {
-      response.writeHead(HttpCode.ServerError)
-      response.end(error.message)
-    })
-    // Intercept the response to rewrite absolute redirects against the base path.
-    this.proxy.on("proxyRes", (response, request: ProxyRequest) => {
-      if (response.headers.location && response.headers.location.startsWith("/") && request.base) {
-        response.headers.location = request.base + response.headers.location
+export const getCookieDomain = (host: string, proxyDomains: string[]): string | undefined => {
+  const idx = host.lastIndexOf(":")
+  host = idx !== -1 ? host.substring(0, idx) : host
+  if (
```

Comment:
Hopefully my above comment explains a bit better, the idea is that if you are visiting `8080.codeserver.com` then with this code the cookie will be set at `codeserver.com` instead of `8080.codeserver.com` so that you have a single cookie and only have to log in once.